File: Leetcode/3484 - Design Spreadsheet/solution.cpp
class Spreadsheet {
private:
    vector<vector<int>> grid;
    pair<int, int> getCell(string cell) {
        int col = cell[0] - 'A';
        int row = stoi(cell.substr(1)) - 1;
        return {row, col};
    }

    int getCellValue(string cell) {
        if (isalpha(cell[0])) {
            auto [row, col] = getCell(cell);
            return grid[row][col];
        } else {
            return stoi(cell);
        }
    }
public:
    Spreadsheet(int rows) {
        grid = vector<vector<int>>(rows, vector<int>(26, 0));
    }
    
    void setCell(string cell, int value) {
        auto [row, col] = getCell(cell);
        grid[row][col] = value;
    }
    
    void resetCell(string cell) {
        auto [row, col] = getCell(cell);
        grid[row][col] = 0;
    }
    
    int getValue(string formula) {
        formula = formula.substr(1);
        size_t plusPos = formula.find('+');
        string cell1 = formula.substr(0, plusPos);
        string cell2 = formula.substr(plusPos + 1);
        return getCellValue(cell1) + getCellValue(cell2);
    }
};

File: Leetcode/3484 - Design Spreadsheet/solution.py
class Spreadsheet:

    def __init__(self, rows: int):
        self.grid: List[List[int]] = [[0] * 26 for _ in range(rows)]
    
    def _get_cell(self, cell: str) -> Tuple[int, int]:
        col = ord(cell[0]) - ord('A')
        row = int(cell[1:]) - 1
        return row, col

    def _get_cell_value(self, cell: str) -> int:
        if cell[0].isalpha():
            row, col = self._get_cell(cell)
            return self.grid[row][col]
        return int(cell)

    def setCell(self, cell: str, value: int) -> None:
        row, col = self._get_cell(cell)
        self.grid[row][col] = value

    def resetCell(self, cell: str) -> None:
        row, col = self._get_cell(cell)
        self.grid[row][col] = 0

    def getValue(self, formula: str) -> int:
        expr = formula[1:]
        plus_pos = expr.index('+')
        cell1 = expr[:plus_pos]
        cell2 = expr[plus_pos + 1:]
        return self._get_cell_value(cell1) + self._get_cell_value(cell2)